### BUILDING ATTRIBUTE ANALYSIS FOR OURAY COUNTY

Author: Bryce Young  
Created: 9/10/2024  
Updated: 9/24/2024

The purpose of this notebook is to document the process of data cleaning and analysis for Ouray County tax assessor data. 

There are two reports that were sent to me by Susie Mayfield, the Ouray County Assessor. I explore these dataframes and clean them according to the project requirements. 

In [3]:
import pandas as pd

df1 = pd.read_csv('BuildingChar.csv')
df2 = pd.read_csv('CreateAccount.csv')

print(df1.shape, df2.shape)

(8555, 36) (5680, 274)


C:\Users\by197116\AppData\Local\Temp\ipykernel_24372\1232649359.py:4: DtypeWarning: Columns (1,10,21,29,37,46,50,58,65) have mixed types. Specify dtype option on import or set low_memory=False.
  df2 = pd.read_csv('CreateAccount.csv')


Note immediately from the warning that there are mixed dtypes in the columns of df2. That is probably because of all the nan values read in by numpy. We're going to bookmark this issue and continue forward for now.

First, let's preview the data.

In [28]:
# Display all cols instead of truncated version
pd.set_option('display.max_columns', None)

df1.head()

Account Number Parcel Number Model Name  Abstract Code  Actual Year Built  \
0        R002405  403514300004  RESIDENCE         1277.0             1985.0   
1        R002405  403514300004       XFOB         1277.0                NaN   
2        R002639  403514300005  RESIDENCE         1277.0             1983.0   
3        R006750  403515200002  RESIDENCE         1277.0             2007.0   
4        R006750  403515200002  RESIDENCE         1277.0             2000.0   

  Air Conditioning Arch Style Base Value  Baths  Bedrooms  ... SPCDUC  SPCDPC  \
0             NONE    LOG KIT         Q4    0.0       1.0  ...    NaN     NaN   
1                0          0          0    NaN       NaN  ...    NaN     NaN   
2             NONE      RANCH      CABIN    0.0       1.0  ...    NaN     NaN   
3                0      RANCH         Q3    2.5       2.0  ...    NaN     NaN   
4                0  ONE STORY         Q4    1.0       1.0  ...    NaN     NaN   

   SPCDPD  Stories               USe     SQFT  Calculated Value Heated Area  \
0     NaN      0.0     AG SF HOUSING   768.00            248110    768.0000   
1     NaN      NaN                 0      NaN              1580         NaN   
2     NaN      0.0     AG SF HOUSING   192.00              7600    192.0000   
3     NaN      NaN  RESIDENTIAL IMPD  1880.87           1092750   1880.8752   
4     NaN      NaN  RESIDENTIAL IMPD   510.78            201890    510.7894   

  Actual Area Reporting Sq Ft  
0   1152.0000          768.00  
1         NaN          144.00  
2    192.0000          192.00  
3   4382.2500         1880.87  
4    692.7892          510.78  

[5 rows x 36 columns]

In [18]:
# Display all cols instead of truncated version
pd.set_option('display.max_columns', None)

df2.head()

Account Number Parcel Number                    Source  Instance  \
0        R002405  403514300004  Residence - non-integral         1   
1        R002639  403514300005  Residence - non-integral         1   
2        R006750  403515200002               Residence 1         1   
3        R006750  403515200002               Residence 1         2   
4        R006750  403515200002               Residence 2         1   

  MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

  MOBILE_HOME: Account Number\n(AccountNumber)  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   MOBILE_HOME: Actual Area\n(ActualArea)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

   MOBILE_HOME: Actual Year Built\n(AYB)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   MOBILE_HOME: ADD - Addition\n(SubArea_ADD)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

   MOBILE_HOME: Adjusted Area\n(AdjustedArea)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

  MOBILE_HOME: Air Conditioning\n(AIRC)  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   

  MOBILE_HOME: Architecture Style\n(ARCH)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

   MOBILE_HOME: Area Acres\n(Area_ACRES)  MOBILE_HOME: Area SQFT\n(Area_SQFT)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                    NaN                                  NaN   
4                                    NaN                                  NaN   

   MOBILE_HOME: Area Units\n(Area_UNITS)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   MOBILE_HOME: BAS - Base\n(SubArea_BAS) MOBILE_HOME: Base Value\n(BVAL)  \
0                                     NaN                             NaN   
1                                     NaN                             NaN   
2                                     NaN                             NaN   
3                                     NaN                             NaN   
4                                     NaN                             NaN   

   MOBILE_HOME: Baths\n(BATH)  MOBILE_HOME: Bedrooms\n(BDRM)  \
0                         NaN                     

In [30]:
df1_cols = (df1.columns).tolist()
df1_cols

['Account Number',
 'Parcel Number',
 'Model Name',
 'Abstract Code',
 'Actual Year Built',
 'Air Conditioning',
 'Arch Style',
 'Base Value',
 'Baths',
 'Bedrooms',
 'Condition',
 'Construction Quality',
 'Economic Obs',
 'Effective Year Built',
 'Fixtures',
 'Frames',
 'Functional Obsolescence',
 'Heating Fuel',
 'Heating Type',
 'Neighborhood',
 'Roof Cover',
 'Roof Structure',
 'Rooms',
 'Depreciation',
 'SPCD',
 'SPCDAP',
 'SPCDUC',
 'SPCDPC',
 'SPCDPD',
 'Stories',
 'USe',
 'SQFT',
 'Calculated Value',
 'Heated Area',
 'Actual Area',
 'Reporting Sq Ft']

In [31]:
df2_cols = (df2.columns).tolist()
df2_cols

['Account Number',
 'Parcel Number',
 'Source',
 'Instance',
 'MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)',
 'MOBILE_HOME: Account Number\n(AccountNumber)',
 'MOBILE_HOME: Actual Area\n(ActualArea)',
 'MOBILE_HOME: Actual Year Built\n(AYB)',
 'MOBILE_HOME: ADD - Addition\n(SubArea_ADD)',
 'MOBILE_HOME: Adjusted Area\n(AdjustedArea)',
 'MOBILE_HOME: Air Conditioning\n(AIRC)',
 'MOBILE_HOME: Architecture Style\n(ARCH)',
 'MOBILE_HOME: Area Acres\n(Area_ACRES)',
 'MOBILE_HOME: Area SQFT\n(Area_SQFT)',
 'MOBILE_HOME: Area Units\n(Area_UNITS)',
 'MOBILE_HOME: BAS - Base\n(SubArea_BAS)',
 'MOBILE_HOME: Base Value\n(BVAL)',
 'MOBILE_HOME: Baths\n(BATH)',
 'MOBILE_HOME: Bedrooms\n(BDRM)',
 'MOBILE_HOME: Category\n(Category)',
 'MOBILE_HOME: Category\n(PublicCategory)',
 'MOBILE_HOME: Ceiling\n(CEIL)',
 'MOBILE_HOME: Condition\n(COND)',
 'MOBILE_HOME: CPF - Carport FL\n(SubArea_CPF)',
 'MOBILE_HOME: CPP - Carport PT\n(SubArea_CPP)',
 'MOBILE_HOME: Depreciation 06\n(DEPR06)',
 'MOBILE_HOME: Eco

In [32]:
print(len(df1_cols), len(df2_cols))

36 274


The "Building Characteristics" sheet has **8555 rows x 36 columns** while the "Create Account" sheet has **5680 rows x 274 columns**. This is because the "Building Characteristics" report has a line item for every building on the account. This requires identifying the primary residence. According to Susie, the "Residence 1" model is the primary, but not always. The residence with the larger SQFT will most likely be the primary residence.

#### NOTES FROM SPRING 2024:

There are a few things I need to do to accommodate this (NOTE: these are not necessarily in order of operation):

1. Identify the columns that I want to keep between both dataframes
2. Group rows by parcel number
3. Number of outbuildings = number of rows of same parcel number, less 1
4. In df1 and df2, I have to determine different criteria by which I decide which row to keep from unique parcel numbers.
    - in df1, the row to keep is the nanmax of 'SQFT'
    - in df2, the row to keep is the nanmax of 'MOBILE_HOME: Area SQFT\n(Area_SQFT)' or 'RESIDENCE: Area SQFT\n(Area_SQFT)' or 'COMMERCIAL: Area SQFT\n(Area_SQFT)'
5. In df2, if 'MOBILE_HOME: Area SQFT\n(Area_SQFT)' has a value in a row, create a new column called 'Type' and label it 'MOBILE' for that row.
    - Do the same thing for 'RESIDENCE' and 'COMMERCIAL'
    - To describe the process for clarity, the objective is to consolidate columns.
    - A new column will be created which identifies the primary structure 'Type'
    - The remainder of the columns will be consolidated. This will be a lot of work unless I figure out something clever

Let's take df2 and extract the suffixes to mobile, residence and commercial columns to explore the possibility of consolidating the dataframe. 

First, I'm going to create sets for each type's column suffixes to compare them. Sets only contain unique values, which shouldn't be an issue. Sets are better for testing for membership and performing operations like union, intersection and difference. That's what we're going to do.

In [7]:
mob_cols = set()
res_cols = set()
com_cols = set()

for col in df2.columns:
    if col.startswith('MOBILE_HOME: '):
        mob_cols.add(col.replace('MOBILE_HOME: ', ''))
    elif col.startswith('RESIDENCE: '):
        res_cols.add(col.replace('RESIDENCE: ', ''))
    elif col.startswith('COMMERCIAL: '):
        com_cols.add(col.replace('COMMERCIAL: ', ''))

print('number of mobile columns: ', len(mob_cols))
print('number of residential columns: ', len(res_cols))
print('number of commercial columns: ', len(com_cols))
print('total columns: ', len(mob_cols) + len(res_cols) + len(com_cols))

number of mobile columns:  71
number of residential columns:  105
number of commercial columns:  94
total columns:  270


Now let's create a set called `uni_cols` which is supposed to stand for universal columns, containing column suffixes that are common across all three lists. 

In [13]:
uni_cols = set(res_cols.intersection(mob_cols, com_cols))

print('number of universal columns: ', len(uni_cols))
uni_cols

number of universal columns:  63


{'Abstract Code\n(ABSTRACT_CODE)',
 'Account Number\n(AccountNumber)',
 'Actual Year Built\n(AYB)',
 'Adjusted Area\n(AdjustedArea)',
 'Air Conditioning\n(AIRC)',
 'Architecture Style\n(ARCH)',
 'Area Acres\n(Area_ACRES)',
 'Area SQFT\n(Area_SQFT)',
 'Area Units\n(Area_UNITS)',
 'BAS - Base\n(SubArea_BAS)',
 'Base Value\n(BVAL)',
 'Baths\n(BATH)',
 'Bedrooms\n(BDRM)',
 'CPF - Carport FL\n(SubArea_CPF)',
 'Category\n(Category)',
 'Category\n(PublicCategory)',
 'Ceiling\n(CEIL)',
 'Condition\n(COND)',
 'Depreciation 06\n(DEPR06)',
 'Economic Obsolescence\n(ECON)',
 'Effective Year Built\n(EYB)',
 'Entry Date\n(EntryDate)',
 'Entry Date\n(PublicEntryDate)',
 'Exterior Percent\n(EXW_PERCENT)',
 'Exterior Wall\n(EXW)',
 'Fixtures\n(FIXT)',
 'Floor\n(FLR)',
 'Floor Percent\n(FLR_PERCENT)',
 'Foundation\n(FOUND)',
 'Frame\n(FRME)',
 'Functional Obsolescence\n(FUNC)',
 'Heated Area\n(HeatedArea)',
 'Heating Fuel\n(HTFL)',
 'Heating Type\n(HTTP)',
 'Height\n(HGHT)',
 'Interior Percent\n(INT_PER

This makes me curious as to which 42 columns are unique to residential, 31 to commercial and 8 to mobile. Just for curiosity's sake, let's find out.

In [14]:

mob_cols_unique = mob_cols.difference(res_cols, com_cols)
res_cols_unique = res_cols.difference(mob_cols, com_cols)
com_cols_unique = com_cols.difference(mob_cols, res_cols)

print("Unique mobile home columns:", len(mob_cols_unique))
print("Unique residence columns:", len(res_cols_unique))
print("Unique commercial columns:", len(com_cols_unique))


Unique mobile home columns: 2
Unique residence columns: 26
Unique commercial columns: 19


The columns I want to keep from residential are 'CPG - Carport Good\n(SubArea_CPG)', 'DGR - Detached Garage\n(SubArea_DGR)', and 'PLS - Pole Shed\n(SubArea_PLS)'.

There are no columns that I care to keep from commercial or mobile units. So I will keep these three columns in addition to the ones in the `uni_cols` set.

Now, let's make a major change to `df2`. We're going to remove the prefixes `'MOBILE_HOME: '`, `'RESIDENCE: '`, and `'COMMERCIAL: '` from the dataframe, drop any columns that are not in `uni_cols`, and create a new column called `'Type'` which will contain the values of the prefixes. This requires a script that will do the following, in order:  
1. Create a new column called `'Type'`
2. Drop columns whose titles end with something that is not an element of the `uni_cols` set.
3. If a 'Mobile' column contains a value in a row, but there is no value in residence or commercial, then add 'mobile' as the value under the `Type` column for which the row contains a value.
4. Raise an error/warning if a row contains values in two or more of mobile, residential, and commercial columns.
5. Do the same thing for Residential and Commercial.

PAUSE AND CHECK FOR ERRORS

6. If errors, clean as necessary. If no errors, proceed to 7
7. Remove prefixes from columns.
8. Merge columns with the same title, making sure real values overwrite no data.

But first I want to make sure that 'CPG - Carport Good\n(SubArea_CPG)', 'DGR - Detached Garage\n(SubArea_DGR)', and 'PLS - Pole Shed\n(SubArea_PLS)' are added to `uni_cols`. 

In [20]:
# Add three cols from the residential list to the uni_cols set
uni_cols.update([
    'CPG - Carport Good\n(SubArea_CPG)',
    'DGR - Detached Garage\n(SubArea_DGR)',
    'PLS - Pole Shed\n(SubArea_PLS)'
])

# QAQC
print('number of universal columns: ', len(uni_cols))
uni_cols

number of universal columns:  66


{'Abstract Code\n(ABSTRACT_CODE)',
 'Account Number\n(AccountNumber)',
 'Actual Year Built\n(AYB)',
 'Adjusted Area\n(AdjustedArea)',
 'Air Conditioning\n(AIRC)',
 'Architecture Style\n(ARCH)',
 'Area Acres\n(Area_ACRES)',
 'Area SQFT\n(Area_SQFT)',
 'Area Units\n(Area_UNITS)',
 'BAS - Base\n(SubArea_BAS)',
 'Base Value\n(BVAL)',
 'Baths\n(BATH)',
 'Bedrooms\n(BDRM)',
 'CPF - Carport FL\n(SubArea_CPF)',
 'CPG - Carport Good\n(SubArea_CPG)',
 'Category\n(Category)',
 'Category\n(PublicCategory)',
 'Ceiling\n(CEIL)',
 'Condition\n(COND)',
 'DGR - Detached Garage\n(SubArea_DGR)',
 'Depreciation 06\n(DEPR06)',
 'Economic Obsolescence\n(ECON)',
 'Effective Year Built\n(EYB)',
 'Entry Date\n(EntryDate)',
 'Entry Date\n(PublicEntryDate)',
 'Exterior Percent\n(EXW_PERCENT)',
 'Exterior Wall\n(EXW)',
 'Fixtures\n(FIXT)',
 'Floor\n(FLR)',
 'Floor Percent\n(FLR_PERCENT)',
 'Foundation\n(FOUND)',
 'Frame\n(FRME)',
 'Functional Obsolescence\n(FUNC)',
 'Heated Area\n(HeatedArea)',
 'Heating Fuel\n(H

In [26]:
# Add type col
df2['Type'] = ''

# Specify columns to keep, including the ones that don't have structuretype prefix
cols2keep = [
 'Account Number',
 'Parcel Number',
 'Source',
 'Instance',
 'Type'
]

for col in df2.columns:
    if any(col.endswith(suffix) for suffix in uni_cols):
        cols2keep.append(col)

# We had 63 common cols, added 3 from residential and 5 miscellaneous. Check that that's the same length as cols2keep
len(cols2keep) == (63*3 + 8)

True

In [27]:
df2 = df2[cols2keep]
df2.head()

Account Number Parcel Number                    Source  Instance Type  \
0        R002405  403514300004  Residence - non-integral         1        
1        R002639  403514300005  Residence - non-integral         1        
2        R006750  403515200002               Residence 1         1        
3        R006750  403515200002               Residence 1         2        
4        R006750  403515200002               Residence 2         1        

  MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

  MOBILE_HOME: Account Number\n(AccountNumber)  \
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   

   MOBILE_HOME: Actual Year Built\n(AYB)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   MOBILE_HOME: Adjusted Area\n(AdjustedArea)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

  MOBILE_HOME: Air Conditioning\n(AIRC)  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   

  MOBILE_HOME: Architecture Style\n(ARCH)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

   MOBILE_HOME: Area Acres\n(Area_ACRES)  MOBILE_HOME: Area SQFT\n(Area_SQFT)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                    NaN                                  NaN   
4                                    NaN                                  NaN   

   MOBILE_HOME: Area Units\n(Area_UNITS)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   MOBILE_HOME: BAS - Base\n(SubArea_BAS) MOBILE_HOME: Base Value\n(BVAL)  \
0                                     NaN                             NaN   
1                                     NaN                             NaN   
2                                     NaN                             NaN   
3                                     NaN                             NaN   
4                                     NaN                             NaN   

   MOBILE_HOME: Baths\n(BATH)  MOBILE_HOME: Bedrooms\n(BDRM)  \
0                         NaN                            NaN   
1                         NaN                            NaN   
2                         NaN                            NaN   
3                         NaN                            NaN   
4                         NaN                            NaN   

   MOBILE_HOME: Category\n(Category)  MOBILE_HOME: Category\n(PublicCategory)  \
0                                NaN                                      NaN   
1                                NaN                                      NaN   
2                     

#### Remove Duplicate Account Numbers

We can see that there are duplicate parcel numbers. Looks like there is a helper column called 'Instance' which tells you how many times that parcel number occurs. We're going to drop rows where 'Instance' > 1. We want to track how many rows were dropped, just out of curiosity.

NOTE: Another thing I could do is count the number of duplicate instances and save this as information in a new column.

In [28]:
multiple_instances = len(df2)

df2 = df2[df2['Instance'] == 1]

print('rows dropped: ', multiple_instances - len(df2))

1395

In [38]:
df2['Parcel Number'].duplicated().any()

np.True_

Looks like we dropped many duplicates but there are still duplicate account numbers and parcel numbers.

In [31]:
df2[df2['Parcel Number'].duplicated(keep=False)]

Account Number Parcel Number                                    Source  \
2           R006750  403515200002                               Residence 1   
4           R006750  403515200002                               Residence 2   
6           R002330  403517400011                               Residence 1   
7           R002330  403517400011                               Residence 2   
8           R002330  403517400011                               Residence 3   
...             ...           ...                                       ...   
5525        R006278  IMPRONLY6278                    Commercial - GATEHOUSE   
5527        R006278  IMPRONLY6278                               Residence 1   
5528        R006278  IMPRONLY6278                               Residence 2   
5536        R006790  IMPRONLY6790                  Residence 1 - apartments   
5537        R006790  IMPRONLY6790  Residence 2 - public space for community   

      Instance Type MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)  \
2            1                                              NaN   
4            1                                              NaN   
6            1                                              NaN   
7            1                                              NaN   
8            1                                              NaN   
...        ...  ...                                         ...   
5525         1                                              NaN   
5527         1                                              NaN   
5528         1                                              NaN   
5536         1                                              NaN   
5537         1                                              NaN   

     MOBILE_HOME: Account Number\n(AccountNumber)  \
2                                             NaN   
4                                             NaN   
6                                             NaN   
7                                             NaN   
8                                             NaN   
...                                           ...   
5525                                          NaN   
5527                                          NaN   
5528                                          NaN   
5536                                          NaN   
5537                                          NaN   

      MOBILE_HOME: Actual Year Built\n(AYB)  \
2                                       NaN   
4                                       NaN   
6                                       NaN   
7                                       NaN   
8                                       NaN   
...                                     ...   
5525                                    NaN   
5527                                    NaN   
5528                                    NaN   
5536                                    NaN   
5537                                    NaN   

      MOBILE_HOME: Adjusted Area\n(AdjustedArea)  \
2                                            NaN   
4                                            NaN   
6                                            NaN   
7                                            NaN   
8                                            NaN   
...                                          ...   
5525                                         NaN   
5527                                         NaN   
5528                                         NaN   
5536                                         NaN   
5537                                         NaN   

     MOBILE_HOME: Air Conditioning\n(AIRC)  \
2                                      NaN   
4                                      NaN   
6                                      NaN   
7                                      NaN   
8                                      NaN   
...                                    ...   
5525                                   NaN   
5527                                   NaN   
5528      

In [33]:
test = df2[df2['Account Number'] == 'R006278']
test

Account Number Parcel Number                  Source  Instance Type  \
5525        R006278  IMPRONLY6278  Commercial - GATEHOUSE         1        
5527        R006278  IMPRONLY6278             Residence 1         1        
5528        R006278  IMPRONLY6278             Residence 2         1        

     MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)  \
5525                                         NaN   
5527                                         NaN   
5528                                         NaN   

     MOBILE_HOME: Account Number\n(AccountNumber)  \
5525                                          NaN   
5527                                          NaN   
5528                                          NaN   

      MOBILE_HOME: Actual Year Built\n(AYB)  \
5525                                    NaN   
5527                                    NaN   
5528                                    NaN   

      MOBILE_HOME: Adjusted Area\n(AdjustedArea)  \
5525                                         NaN   
5527                                         NaN   
5528                                         NaN   

     MOBILE_HOME: Air Conditioning\n(AIRC)  \
5525                                   NaN   
5527                                   NaN   
5528                                   NaN   

     MOBILE_HOME: Architecture Style\n(ARCH)  \
5525                                     NaN   
5527                                     NaN   
5528                                     NaN   

      MOBILE_HOME: Area Acres\n(Area_ACRES)  \
5525                                    NaN   
5527                                    NaN   
5528                                    NaN   

      MOBILE_HOME: Area SQFT\n(Area_SQFT)  \
5525                                  NaN   
5527                                  NaN   
5528                                  NaN   

      MOBILE_HOME: Area Units\n(Area_UNITS)  \
5525                                    NaN   
5527                                    NaN   
5528                                    NaN   

      MOBILE_HOME: BAS - Base\n(SubArea_BAS) MOBILE_HOME: Base Value\n(BVAL)  \
5525                                     NaN                             NaN   
5527                                     NaN                             NaN   
5528                                     NaN                             NaN   

      MOBILE_HOME: Baths\n(BATH)  MOBILE_HOME: Bedrooms\n(BDRM)  \
5525                         NaN                            NaN   
5527                         NaN                            NaN   
5528                         NaN                            NaN   

      MOBILE_HOME: Category\n(Category)  \
5525                                NaN   
5527                                NaN   
5528                                NaN   

      MOBILE_HOME: Category\n(PublicCategory) MOBILE_HOME: Ceiling\n(CEIL)  \
5525                                      NaN                          NaN   
5527                                      NaN                          NaN   
5528                                      NaN                          NaN   

     MOBILE_HOME: Condition\n(COND)  \
5525                            NaN   
5527                            NaN   
5528                            NaN   

      MOBILE_HOME: CPF - Carport FL\n(SubArea_CPF)  \
5525                                           NaN   
5527                                           NaN   
5528                                           NaN   

     MOBILE_HOME: Depreciation 06\n(DEPR06)  \
5525                                    NaN   
5527                                    NaN   
5528                                    NaN   

      MOBILE_HOME: Economic Obsolescence\n(ECON)  \
5525                                         NaN   
5527                                         NaN   
5528                                         NaN   

      MOBILE_HOME: Effective Year Built\n(EYB)  \
5525                                       Na

In [34]:
df2 = df2[~df2['Parcel Number'].str.startswith('IMPRONLY', na=False)]
df2.shape

(4270, 197)

In [44]:
df2['Account Number'].duplicated().any()

np.True_

Check if there are rows with values in columns with multiple prefixes.

In [40]:
# Create lists of mob, res, and com columns
mob_cols = [col for col in df2.columns if col.startswith('MOB')]
res_cols = [col for col in df2.columns if col.startswith('RES')]
com_cols = [col for col in df2.columns if col.startswith('COM')]

# Create mask where each row contains values in two or more of these sets of columns
mask = (
    (df2[mob_cols].notna().sum(axis=1) > 0).astype(int) + 
    (df2[res_cols].notna().sum(axis=1) > 0).astype(int) + 
    (df2[com_cols].notna().sum(axis=1) > 0).astype(int)
) >= 2

rows_with_multiple_prefixes = df2[mask]
rows_with_multiple_prefixes.shape

(0, 197)

Good, there are zero rows with multiple prefixes. Check if there are account numbers with values in columns with multiple prefixes.

In [43]:
df2['Prefix Count'] = (
    (df2[mob_cols].notna().sum(axis=1) > 0).astype(int) + 
    (df2[res_cols].notna().sum(axis=1) > 0).astype(int) + 
    (df2[com_cols].notna().sum(axis=1) > 0).astype(int)
)

multiple_prefix_accounts = df2.groupby('Parcel Number')['Prefix Count'].max()

multiple_prefix_accounts = multiple_prefix_accounts[multiple_prefix_accounts >= 2]

print(multiple_prefix_accounts)

Series([], Name: Prefix Count, dtype: int64)


There are zero account numbers or parcel numbers with values in multiple prefixes. However, there are still duplicate account numbers and parcel numbers.

In [46]:
# Check for duplicate mobile accounts
mobile_accounts = df2[df2['Account Number'].str.startswith('M', na=False)]
mobile_accounts['Account Number'].duplicated().any()

np.False_

In [54]:
# Remove rows where 'Source' ends with '2'
df2 = df2[~df2['Source'].str.endswith('2', na=False)]
df2['Account Number'].duplicated().any()

np.True_

In [58]:
duplicate_accounts = df2[df2['Account Number'].duplicated(keep=False)]
duplicate_accounts

Account Number Parcel Number                           Source  Instance  \
6           R002330  403517400011                      Residence 1         1   
8           R002330  403517400011                      Residence 3         1   
9           R002330  403517400011                      Residence 4         1   
11          R002607  403715100008       Residence 1 - Manufactured         1   
12          R002607  403715100008               Residence 2 - Camp         1   
...             ...           ...                              ...       ...   
5497        R001711  456907300003     Residence 1 - Main Res Cabin         1   
5498        R001711  456907300003         Residence 2 - Bath House         1   
5499        R001711  456907300003       Residence 3 - Guest Cabin          1   
5515        R006661  477714100002           Commercial - Lodge B&B         1   
5516        R006661  477714100002  Residence - Residential Portion         1   

     Type MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)  \
6                                                 NaN   
8                                                 NaN   
9                                                 NaN   
11                                                NaN   
12                                                NaN   
...   ...                                         ...   
5497                                              NaN   
5498                                              NaN   
5499                                              NaN   
5515                                              NaN   
5516                                              NaN   

     MOBILE_HOME: Account Number\n(AccountNumber)  \
6                                             NaN   
8                                             NaN   
9                                             NaN   
11                                            NaN   
12                                            NaN   
...                                           ...   
5497                                          NaN   
5498                                          NaN   
5499                                          NaN   
5515                                          NaN   
5516                                          NaN   

      MOBILE_HOME: Actual Year Built\n(AYB)  \
6                                       NaN   
8                                       NaN   
9                                       NaN   
11                                      NaN   
12                                      NaN   
...                                     ...   
5497                                    NaN   
5498                                    NaN   
5499                                    NaN   
5515                                    NaN   
5516                                    NaN   

      MOBILE_HOME: Adjusted Area\n(AdjustedArea)  \
6                                            NaN   
8                                            NaN   
9                                            NaN   
11                                           NaN   
12                                           NaN   
...                                          ...   
5497                                         NaN   
5498                                         NaN   
5499                                         NaN   
5515                                         NaN   
5516                                         NaN   

     MOBILE_HOME: Air Conditioning\n(AIRC)  \
6                                      NaN   
8                                      NaN   
9                                      NaN   
11                                     NaN   
12                                     NaN   
...                                    ...   
5497                                   NaN   
5498                                   NaN   
5499                                   NaN   
5515                                   NaN   
5516                      

In [59]:
df2 = df2[~df2['Source'].str.contains(r'[2-9]', na=False)]
duplicate_accounts = df2[df2['Account Number'].duplicated(keep=False)]
duplicate_accounts

Account Number Parcel Number                           Source  Instance  \
32          R000355  403722300015          Commercial 1 - Cabin #1         1   
39          R000355  403722300015  Residence - Caretaker Residence         1   
151         R004141  403923100045                Commercial - Shop         1   
152         R004141  403923100045                        Residence         1   
333         R004128  404117300007                       Commercial         1   
...             ...           ...                              ...       ...   
5073        R001586  451711200002                        Residence         1   
5476        R001200  456726200008               Commercial - LODGE         1   
5477        R001200  456726200008            Residence - RESIDENCE         1   
5515        R006661  477714100002           Commercial - Lodge B&B         1   
5516        R006661  477714100002  Residence - Residential Portion         1   

     Type MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)  \
32                                                NaN   
39                                                NaN   
151                                               NaN   
152                                               NaN   
333                                               NaN   
...   ...                                         ...   
5073                                              NaN   
5476                                              NaN   
5477                                              NaN   
5515                                              NaN   
5516                                              NaN   

     MOBILE_HOME: Account Number\n(AccountNumber)  \
32                                            NaN   
39                                            NaN   
151                                           NaN   
152                                           NaN   
333                                           NaN   
...                                           ...   
5073                                          NaN   
5476                                          NaN   
5477                                          NaN   
5515                                          NaN   
5516                                          NaN   

      MOBILE_HOME: Actual Year Built\n(AYB)  \
32                                      NaN   
39                                      NaN   
151                                     NaN   
152                                     NaN   
333                                     NaN   
...                                     ...   
5073                                    NaN   
5476                                    NaN   
5477                                    NaN   
5515                                    NaN   
5516                                    NaN   

      MOBILE_HOME: Adjusted Area\n(AdjustedArea)  \
32                                           NaN   
39                                           NaN   
151                                          NaN   
152                                          NaN   
333                                          NaN   
...                                          ...   
5073                                         NaN   
5476                                         NaN   
5477                                         NaN   
5515                                         NaN   
5516                                         NaN   

     MOBILE_HOME: Air Conditioning\n(AIRC)  \
32                                     NaN   
39                                     NaN   
151                                    NaN   
152                                    NaN   
333                                    NaN   
...                                    ...   
5073                                   NaN   
5476                                   NaN   
5477                                   NaN   
5515                                   NaN   
5516                      

Getting craftier, I want to drop rows in df2 where - if the account number has a duplicate - drop the row with the smaller value in the 'RESIDENCE: Area SQFT\n(Area_SQFT)' column. Please note that I want to treat NaN as the smallest value.

In [61]:
# Step 1: Sort the DataFrame by 'Account Number' and 'RESIDENCE: Area SQFT\n(Area_SQFT)'
# NaN values are considered smallest when using ascending=False
df2_sorted = df2.sort_values(by=['Account Number', 'RESIDENCE: Area SQFT\n(Area_SQFT)'], ascending=[True, False], na_position='last')

# Step 2: Drop duplicates based on 'Account Number', keeping the first (which will be the largest 'Area SQFT')
df2_v2 = df2_sorted.drop_duplicates(subset='Account Number', keep='first')

In [62]:
df2_v2['Account Number'].duplicated().any()

np.False_

Now we no longer have duplicate account numbers. Let's see how many rows there are in this df. 

In [64]:
df2_v2.columns

Index(['Account Number', 'Parcel Number', 'Source', 'Instance', 'Type',
       'MOBILE_HOME: Abstract Code\n(ABSTRACT_CODE)',
       'MOBILE_HOME: Account Number\n(AccountNumber)',
       'MOBILE_HOME: Actual Year Built\n(AYB)',
       'MOBILE_HOME: Adjusted Area\n(AdjustedArea)',
       'MOBILE_HOME: Air Conditioning\n(AIRC)',
       ...
       'COMMERCIAL: SPCDAP\n(SPCDAP)',
       'COMMERCIAL: STG - Storage\n(SubArea_STG)', 'COMMERCIAL: STYS\n(STYS)',
       'COMMERCIAL: TWO - Second Floor\n(SubArea_TWO)',
       'COMMERCIAL: Use\n(USE)', 'COMMERCIAL: Value 98\n(VALUE_98)',
       'COMMERCIAL: Value 99\n(VALUE_99)',
       'COMMERCIAL: Version Start\n(VersionStart)',
       'COMMERCIAL: WBL - Wood Balcony\n(SubArea_WBL)', 'Prefix Count'],
      dtype='object', length=198)

#### Fill in 'Type' column

Now, getting even craftier, we have to determine which structure type (com, res, mob) an account number is based on which columns it has values in. If any of mob_cols contain values in a row, we're going to assign 'Mobile Home' to that row's 'Type'. Same with com_cols and res_cols. Recall we have already made these lists.

In [65]:
# Define function to do this
def determine_type(row):
    if row[mob_cols].notna().any():
        return 'Mobile Home'
    elif row[res_cols].notna().any():
        return 'Residence'
    elif row[com_cols].notna().any():
        return 'Commercial'
    else:
        return ''     # Value will be empty if all values in the row are NaN

In [66]:
df2_v2['Type'] = df2_v2.apply(determine_type, axis=1)
(df2_v2['Type'] == '').any()

C:\Users\by197116\AppData\Local\Temp\ipykernel_24372\3627846734.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2_v2['Type'] = df2_v2.apply(determine_type, axis=1)


np.False_

#### Remove column prefixes and squeeze dataframe

The final craftiness is to remove the prefixes of the columns, and merge columns with identical names so that non-NaN values overwrite NaN.

In [69]:
df2_v2.columns = df2_v2.columns.str.replace(r'^(MOBILE_HOME: |RESIDENCE: |COMMERCIAL: )', '', regex=True)

df2_clean = df2_v2.groupby(axis=1, level=0).first()

df2_clean.shape

C:\Users\by197116\AppData\Local\Temp\ipykernel_24372\2278412523.py:3: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.
  df2_clean = df2_v2.groupby(axis=1, level=0).first()


(3480, 72)

In [70]:
df2_clean.head()

Abstract Code\n(ABSTRACT_CODE) Account Number  \
5538  9292 - EXEMPT-TITLED MOBILE HOMES $28K OR LESS...        M000034   
5539       1235 - RES-MANUFACTURED HOUSING-IMPROVEMENTS        M000037   
5540  9292 - EXEMPT-TITLED MOBILE HOMES $28K OR LESS...        M000040   
5541       1235 - RES-MANUFACTURED HOUSING-IMPROVEMENTS        M000042   
5542       1235 - RES-MANUFACTURED HOUSING-IMPROVEMENTS        M000043   

     Account Number\n(AccountNumber)  Actual Year Built\n(AYB)  \
5538                         M000034                    1976.0   
5539                         M000037                    1992.0   
5540                         M000040                    1985.0   
5541                         M000042                    2002.0   
5542                         M000043                    1975.0   

      Adjusted Area\n(AdjustedArea) Air Conditioning\n(AIRC)  \
5538                        1157.55                  0 - N/A   
5539                         908.80                  0 - N/A   
5540                        1126.40                  0 - N/A   
5541                        1318.60                  0 - N/A   
5542                        1097.60                  0 - N/A   

     Architecture Style\n(ARCH)  Area Acres\n(Area_ACRES)  \
5538                0 - PRE-HUD                       0.0   
5539             1 - SINGLE                           0.0   
5540             1 - SINGLE                           0.0   
5541             1 - SINGLE                           0.0   
5542                0 - PRE-HUD                       0.0   

      Area SQFT\n(Area_SQFT)  Area Units\n(Area_UNITS)  \
5538                  1064.0                       0.0   
5539                   840.0                       0.0   
5540                   980.0                       0.0   
5541                  1216.0                       0.0   
5542                  1064.0                       0.0   

      BAS - Base\n(SubArea_BAS)         Base Value\n(BVAL)  Baths\n(BATH)  \
5538                     1064.0  123500 - Mobile Park BVAL            2.0   
5539                      840.0  123500 - Mobile Park BVAL            0.0   
5540                      980.0  123500 - Mobile Park BVAL            NaN   
5541                     1216.0  123500 - Mobile Park BVAL            1.0   
5542                     1064.0  123500 - Mobile Park BVAL            0.0   

      Bedrooms\n(BDRM)  CPF - Carport FL\n(SubArea_CPF)  \
5538               3.0                            231.0   
5539               0.0                              NaN   
5540               NaN                              NaN   
5541               2.0                              NaN   
5542               0.0                              NaN   

      CPG - Carport Good\n(SubArea_CPG) Category\n(Category)  \
5538                                NaN                 None   
5539                                NaN                 None   
5540                                NaN                 None   
5541                                NaN                 None   
5542                                NaN                 None   

     Category\n(PublicCategory) Ceiling\n(CEIL)   Condition\n(COND)  \
5538                       None  0 - N/A         5 - BELOW AVG - C5   
5539                       None  0 - N/A              3 - GOOD - C3   
5540                       None  0 - N/A               4 - AVG - C4   
5541                       None  0 - N/A               4 - AVG - C4   
5542                       None  0 - N/A               4 - AVG - C4   

      DGR - Detached Garage\n(SubArea_DGR) Depreciation 06\n(DEPR06)  \
5538                                   NaN             1990 - DEPR06   
5539                                   NaN             2005 - DEPR06   
5540                                   NaN             1985 - DEPR06   
5541                                   NaN             2005 - DEPR06   
5542                                   NaN             1995 - DEPR06   

      Economic Obso

In [71]:
# Save as csv
df2_clean.to_csv('assessor_clean_1.csv', index=False)

## Second dataframe

Now let's take a look at the other dataframe - df1. We're going to explore its data content, and its ability to join with df2_clean (e.g. via Account Number etc.). If there is value, then we will clean it. If there is not, then we will scrap it. 

In [76]:
df1.head()

Account Number Parcel Number Model Name  Abstract Code  Actual Year Built  \
0        R002405  403514300004  RESIDENCE         1277.0             1985.0   
1        R002405  403514300004       XFOB         1277.0                NaN   
2        R002639  403514300005  RESIDENCE         1277.0             1983.0   
3        R006750  403515200002  RESIDENCE         1277.0             2007.0   
4        R006750  403515200002  RESIDENCE         1277.0             2000.0   

  Air Conditioning Arch Style Base Value  Baths  Bedrooms Condition  \
0             NONE    LOG KIT         Q4    0.0       1.0       C-3   
1                0          0          0    NaN       NaN         0   
2             NONE      RANCH      CABIN    0.0       1.0       C-5   
3                0      RANCH         Q3    2.5       2.0       C-3   
4                0  ONE STORY         Q4    1.0       1.0       C-4   

   Construction Quality  Economic Obs  Effective Year Built  Fixtures  \
0                   4.0           0.0                2000.0       0.0   
1                   NaN           NaN                   NaN       NaN   
2                   4.0           0.0                2000.0       0.0   
3                   NaN           NaN                2007.0       NaN   
4                   NaN           NaN                2000.0       NaN   

       Frames  Functional Obsolescence Heating Fuel Heating Type Neighborhood  \
0  LOG                             0.0   WOOD              STOVE    Outlying4   
1           0                      NaN            0            0            0   
2  WOOD                            0.0   NONE               NONE    Outlying4   
3           0                      NaN   PROPANE      FORCED AIR    Outlying4   
4           0                      NaN   GAS               STOVE    Outlying4   

   Roof Cover Roof Structure  Rooms  Depreciation  SPCD  SPCDAP  SPCDUC  \
0  ASPH SHNGL     GABEL STP     0.0        2000.0   NaN     NaN     NaN   
1           0              0    NaN           NaN   NaN     NaN     NaN   
2  GALVANIZED     GABEL STP     0.0        2000.0   NaN     NaN     NaN   
3  ASPH SHNGL     GABEL MED     NaN        2007.0   NaN     NaN     NaN   
4   PRO PANEL     GABEL LOW     NaN        2010.0   NaN     NaN     NaN   

   SPCDPC  SPCDPD  Stories               USe     SQFT  Calculated Value  \
0     NaN     NaN      0.0     AG SF HOUSING   768.00            248110   
1     NaN     NaN      NaN                 0      NaN              1580   
2     NaN     NaN      0.0     AG SF HOUSING   192.00              7600   
3     NaN     NaN      NaN  RESIDENTIAL IMPD  1880.87           1092750   
4     NaN     NaN      NaN  RESIDENTIAL IMPD   510.78            201890   

   Heated Area  Actual Area  Reporting Sq Ft  
0     768.0000    1152.0000           768.00  
1          NaN          NaN           144.00  
2     192.0000     192.0000           192.00  
3    1880.8752    4382.2500          1880.87  
4     510.7894     692.7892           510.78

In [77]:
df1.shape

(8555, 36)

We can see that this dataframe is much larger. Let's see if there are account numbers in df1 that aren't in df2.

In [78]:
accts_only_in_df1 = df1[~df1['Account Number'].isin(df2_clean['Account Number'])]
accts_only_in_df1.shape

(763, 36)

Out of 8555 rows, there are only 763 account numbers in df1 that are not in df2.

How many of those are duplicates?

In [80]:
accts_only_in_df1['Account Number'].duplicated().sum()

np.int64(472)

There are less than 300 structures that are unique to df1. The juice is probably not worth the squeeze to clean this dataframe and merge it to df2.

#### Back to the First Dataframe

Since the second dataframe is probably not worth the effort of extracting an additional 300 structures, we're going to clean up the first one and call it good. If I find out that I have a lot of risk assessments for homes that are not in the first dataframe, then I may come back to this second one in order to get more ground truth observations for my dataset.

To finish cleaning the first dataframe, I'm going to remove columns that I may not care about. These decisions are based on extensive literature review.

In [81]:
import pandas as pd

df = pd.read_csv('assessor_clean_1.csv')
print(df.shape)
df.head()

(3480, 72)


Abstract Code\n(ABSTRACT_CODE) Account Number  \
0  9292 - EXEMPT-TITLED MOBILE HOMES $28K OR LESS...        M000034   
1       1235 - RES-MANUFACTURED HOUSING-IMPROVEMENTS        M000037   
2  9292 - EXEMPT-TITLED MOBILE HOMES $28K OR LESS...        M000040   
3       1235 - RES-MANUFACTURED HOUSING-IMPROVEMENTS        M000042   
4       1235 - RES-MANUFACTURED HOUSING-IMPROVEMENTS        M000043   

  Account Number\n(AccountNumber)  Actual Year Built\n(AYB)  \
0                         M000034                    1976.0   
1                         M000037                    1992.0   
2                         M000040                    1985.0   
3                         M000042                    2002.0   
4                         M000043                    1975.0   

   Adjusted Area\n(AdjustedArea) Air Conditioning\n(AIRC)  \
0                        1157.55                  0 - N/A   
1                         908.80                  0 - N/A   
2                        1126.40                  0 - N/A   
3                        1318.60                  0 - N/A   
4                        1097.60                  0 - N/A   

  Architecture Style\n(ARCH)  Area Acres\n(Area_ACRES)  \
0                0 - PRE-HUD                       0.0   
1             1 - SINGLE                           0.0   
2             1 - SINGLE                           0.0   
3             1 - SINGLE                           0.0   
4                0 - PRE-HUD                       0.0   

   Area SQFT\n(Area_SQFT)  Area Units\n(Area_UNITS)  \
0                  1064.0                       0.0   
1                   840.0                       0.0   
2                   980.0                       0.0   
3                  1216.0                       0.0   
4                  1064.0                       0.0   

   BAS - Base\n(SubArea_BAS)         Base Value\n(BVAL)  Baths\n(BATH)  \
0                     1064.0  123500 - Mobile Park BVAL            2.0   
1                      840.0  123500 - Mobile Park BVAL            0.0   
2                      980.0  123500 - Mobile Park BVAL            NaN   
3                     1216.0  123500 - Mobile Park BVAL            1.0   
4                     1064.0  123500 - Mobile Park BVAL            0.0   

   Bedrooms\n(BDRM)  CPF - Carport FL\n(SubArea_CPF)  \
0               3.0                            231.0   
1               0.0                              NaN   
2               NaN                              NaN   
3               2.0                              NaN   
4               0.0                              NaN   

   CPG - Carport Good\n(SubArea_CPG)  Category\n(Category)  \
0                                NaN                   NaN   
1                                NaN                   NaN   
2                                NaN                   NaN   
3                                NaN                   NaN   
4                                NaN                   NaN   

   Category\n(PublicCategory) Ceiling\n(CEIL)   Condition\n(COND)  \
0                         NaN  0 - N/A         5 - BELOW AVG - C5   
1                         NaN  0 - N/A              3 - GOOD - C3   
2                         NaN  0 - N/A               4 - AVG - C4   
3                         NaN  0 - N/A               4 - AVG - C4   
4                         NaN  0 - N/A               4 - AVG - C4   

   DGR - Detached Garage\n(SubArea_DGR) Depreciation 06\n(DEPR06)  \
0                                   NaN             1990 - DEPR06   
1                                   NaN             2005 - DEPR06   
2                                   NaN             1985 - DEPR06   
3                                   NaN             2005 - DEPR06   
4                                   NaN             1995 - DEPR06   

   Economic Obsolescence\n(ECON)  Effective Year Built\n(EYB)  \
0                            0.0                       1995.0   
1                            0.0                      

I am using the following code block to print unique values of columns that I don't think have a lot of explanatory power, or that I do not understand. If they don't seem to contain a lot of information (For example 'Category\n(Category)' only contains NaN values), then I am adding them to the list below: `cols2drop`

In [120]:
print(df['Version Start\n(VersionStart)'].unique())

[1.66102e+12 1.69309e+12 1.68599e+12 1.68178e+12 1.67717e+12 1.68413e+12
 1.70409e+12 1.68738e+12 1.67832e+12 1.67256e+12 1.68179e+12 1.69143e+12
 1.68625e+12 1.64102e+12 1.68183e+12 1.67872e+12 1.68798e+12 1.69110e+12
 1.70314e+12 1.70621e+12 1.68660e+12 1.67589e+12 1.68676e+12 1.67330e+12
 1.69284e+12 1.68436e+12 1.68548e+12 1.68494e+12 1.68747e+12 1.67822e+12
 1.68019e+12 1.68054e+12 1.68056e+12 1.67415e+12 1.67416e+12 1.68701e+12
 1.67417e+12 1.67338e+12 1.67398e+12 1.67460e+12 1.70664e+12 1.67408e+12
 1.67952e+12 1.67339e+12 1.68011e+12 1.67336e+12 1.67632e+12 1.67469e+12
 1.68608e+12 1.67397e+12 1.67520e+12 1.70597e+12 1.67890e+12 1.67476e+12
 1.67761e+12 1.67535e+12 1.67993e+12 1.67641e+12 1.68175e+12 1.67760e+12
 1.68305e+12 1.68427e+12 1.68564e+12 1.67510e+12 1.67475e+12 1.67891e+12
 1.67892e+12 1.67882e+12 1.68012e+12 1.68177e+12 1.67406e+12 1.70672e+12
 1.67329e+12 1.68772e+12 1.68659e+12 1.68563e+12 1.68496e+12 1.67941e+12
 1.69310e+12 1.67769e+12 1.68435e+12 1.68746e+12 1.

In [122]:
cols2drop = ['Abstract Code\n(ABSTRACT_CODE)',
             'Account Number\n(AccountNumber)',
             'Adjusted Area\n(AdjustedArea)',
             'Area Units\n(Area_UNITS)',
             'BAS - Base\n(SubArea_BAS)',
             'Base Value\n(BVAL)',
             'Baths\n(BATH)',
             'Bedrooms\n(BDRM)',
             'Ceiling\n(CEIL)',
             'Category\n(PublicCategory)',
             'Category\n(Category)',
             'CPG - Carport Good\n(SubArea_CPG)',
             'CPF - Carport FL\n(SubArea_CPF)',
             'Depreciation 06\n(DEPR06)',
             'Economic Obsolescence\n(ECON)',
             'Entry Date\n(EntryDate)',
             'Entry Date\n(PublicEntryDate)',
             'Fixtures\n(FIXT)',
             'Floor Percent\n(FLR_PERCENT)',
             'Functional Obsolescence\n(FUNC)',
             'Heated Area\n(HeatedArea)',
             'Heating Type\n(HTTP)',
             'Instance',
             'Interior Percent\n(INT_PERCENT)',
             'Model\n(Model)',
             'Model\n(publicModel)',
             'Office\n(Office)',
             'Office\n(PublicOffice)',
             'PLS - Pole Shed\n(SubArea_PLS)',
             'PTR - Pointer\n(SubArea_PTR)',
             'Part\n(PART)',
             'Plumbing\n(PLUM)',
             'Prefix Count',
             'Prime Area\n(PrimeArea)',
             'Quality\n(QUAL)',
             'Remarks\n(PublicRemarks)',
             'Remarks\n(Remarks)',
             'Source',
             'SPCDAP\n(SPCDAP)',
             'STG - Storage\n(SubArea_STG)',
             'STYS\n(STYS)',
             'TWO - Second Floor\n(SubArea_TWO)',
             'Use\n(USE)',
             'Value 98\n(VALUE_98)',
             'Value 99\n(VALUE_99)',
             'Version Start\n(VersionStart)'
             ]

print(len(cols2drop))

41


In [123]:
# Drop cols from the cols2drop list
df.drop(columns=cols2drop, inplace=True)
print(df.shape)
df.head()

(3480, 31)


Account Number  Actual Year Built\n(AYB)  Adjusted Area\n(AdjustedArea)  \
0        M000034                    1976.0                        1157.55   
1        M000037                    1992.0                         908.80   
2        M000040                    1985.0                        1126.40   
3        M000042                    2002.0                        1318.60   
4        M000043                    1975.0                        1097.60   

  Air Conditioning\n(AIRC) Architecture Style\n(ARCH)  \
0                  0 - N/A                0 - PRE-HUD   
1                  0 - N/A             1 - SINGLE       
2                  0 - N/A             1 - SINGLE       
3                  0 - N/A             1 - SINGLE       
4                  0 - N/A                0 - PRE-HUD   

   Area Acres\n(Area_ACRES)  Area SQFT\n(Area_SQFT)   Condition\n(COND)  \
0                       0.0                  1064.0  5 - BELOW AVG - C5   
1                       0.0                   840.0       3 - GOOD - C3   
2                       0.0                   980.0        4 - AVG - C4   
3                       0.0                  1216.0        4 - AVG - C4   
4                       0.0                  1064.0        4 - AVG - C4   

   DGR - Detached Garage\n(SubArea_DGR)  Effective Year Built\n(EYB)  \
0                                   NaN                       1995.0   
1                                   NaN                       2005.0   
2                                   NaN                       1985.0   
3                                   NaN                       2005.0   
4                                   NaN                       1995.0   

   Exterior Percent\n(EXW_PERCENT)  Exterior Wall\n(EXW)    Floor\n(FLR)  \
0                            100.0             22 - ALUM  1 - WDJST PLYW   
1                            100.0             4 - FRAME  1 - WDJST PLYW   
2                            100.0             4 - FRAME  1 - WDJST PLYW   
3                            100.0  3 - DESTRESSED METAL  1 - WDJST PLYW   
4                            100.0  3 - DESTRESSED METAL  1 - WDJST PLYW   

   Foundation\n(FOUND)   Frame\n(FRME) Heating Fuel\n(HTFL) Height\n(HGHT)  \
0                  0.0  2 - FRME 2X4         3 - GAS                   NaN   
1                  NaN  3 - FRME 2X6         3 - GAS                   NaN   
2                  NaN  2 - FRME 2X4         3 - GAS                   NaN   
3                  0.0  2 - FRME 2X4         3 - GAS                   NaN   
4                  0.0  2 - FRME 2X4         3 - GAS                   NaN   

   Instance Interior Wall\n(INT)             Neighborhood\n(NBHD)  \
0         1       0 - N/A                    1000200 - 4J MH PARK   
1         1       1 - DRYWALL                1000200 - 4J MH PARK   
2         1       2 - PLSTR/LTH   1000100 - SWISS VILLAGE MH PARK   
3         1       1 - DRYWALL     1000100 - SWISS VILLAGE MH PARK   
4         1       0 - N/A                    1000200 - 4J MH PARK   

   OPP - Open Porch\n(SubArea_OPP)  PTO - Patio\n(SubArea_PTO) Parcel Number  \
0                            176.0                         NaN  MOBILEM00034   
1                            160.0                         NaN  MOBILEM00037   
2                            104.0                         NaN  MOBILEM00040   
3                            212.0                         NaN  MOBILEM00042   
4                             24.0                         NaN  MOBILEM00043   

   Prefix Count Quality\n(QUAL)  RMS\n(RMS) Roof Cover\n(RCVR)  \
0             1  1 - LOW                0.0     3 - METAL        
1             1  2 - FAIR               0.0     3 - METAL        
2             1  2 - FAIR               NaN     3 - METAL        
3             1  2 - FAIR               0.0     3 - METAL        
4             1  1 - LOW                0.0     3 - METAL        

  Roof Structure\n(RSTR)       Source         Type  \
0         4 - ARCH        Mobile Home  

In [141]:
df.to_csv('assessor_clean_v2.csv')

#### Final step: filling NaN values

There is a little data exploration to be done here, but basically we want to avoid having NaN values in this dataframe. In some instances such as 'PTO - Patio\n(SubArea_PTO)' we can confidently replace NaN with 0, because it is a square footage. In other cases, we will have to make judgement calls, possibly ending with dropping rows. 

In [126]:
cols2fillWith0 = [
    'Area Acres\n(Area_ACRES)', 
    'Area SQFT\n(Area_SQFT)', 
    'DGR - Detached Garage\n(SubArea_DGR)', 
    'Foundation\n(FOUND)', 
    'OPP - Open Porch\n(SubArea_OPP)',
    'PTO - Patio\n(SubArea_PTO)',
    'WBL - Wood Balcony\n(SubArea_WBL)'
    ]

df[cols2fillWith0] = df[cols2fillWith0].fillna(0)

Now I am going to headhunt for NaN values column by column, and fill accordingly. 

NOTE: the most up to date CSV is `assessor_clean_v2.csv`


In [131]:
# Determine which columns contain NA values
colsWithNa = []

for col in df.columns:
    if df[col].isna().any():
        colsWithNa.append(col)
    else:
        pass

colsWithNa


['Actual Year Built\n(AYB)',
 'Air Conditioning\n(AIRC)',
 'Architecture Style\n(ARCH)',
 'Condition\n(COND)',
 'Effective Year Built\n(EYB)',
 'Exterior Percent\n(EXW_PERCENT)',
 'Exterior Wall\n(EXW)',
 'Floor\n(FLR)',
 'Frame\n(FRME)',
 'Heating Fuel\n(HTFL)',
 'Height\n(HGHT)',
 'Interior Wall\n(INT)',
 'Neighborhood\n(NBHD)',
 'Quality\n(QUAL)',
 'RMS\n(RMS)',
 'Roof Cover\n(RCVR)',
 'Roof Structure\n(RSTR)']

In [127]:
df.head()

Account Number  Actual Year Built\n(AYB) Air Conditioning\n(AIRC)  \
0        M000034                    1976.0                  0 - N/A   
1        M000037                    1992.0                  0 - N/A   
2        M000040                    1985.0                  0 - N/A   
3        M000042                    2002.0                  0 - N/A   
4        M000043                    1975.0                  0 - N/A   

  Architecture Style\n(ARCH)  Area Acres\n(Area_ACRES)  \
0                0 - PRE-HUD                       0.0   
1             1 - SINGLE                           0.0   
2             1 - SINGLE                           0.0   
3             1 - SINGLE                           0.0   
4                0 - PRE-HUD                       0.0   

   Area SQFT\n(Area_SQFT)   Condition\n(COND)  \
0                  1064.0  5 - BELOW AVG - C5   
1                   840.0       3 - GOOD - C3   
2                   980.0        4 - AVG - C4   
3                  1216.0        4 - AVG - C4   
4                  1064.0        4 - AVG - C4   

   DGR - Detached Garage\n(SubArea_DGR)  Effective Year Built\n(EYB)  \
0                                   0.0                       1995.0   
1                                   0.0                       2005.0   
2                                   0.0                       1985.0   
3                                   0.0                       2005.0   
4                                   0.0                       1995.0   

   Exterior Percent\n(EXW_PERCENT)  Exterior Wall\n(EXW)    Floor\n(FLR)  \
0                            100.0             22 - ALUM  1 - WDJST PLYW   
1                            100.0             4 - FRAME  1 - WDJST PLYW   
2                            100.0             4 - FRAME  1 - WDJST PLYW   
3                            100.0  3 - DESTRESSED METAL  1 - WDJST PLYW   
4                            100.0  3 - DESTRESSED METAL  1 - WDJST PLYW   

   Foundation\n(FOUND)   Frame\n(FRME) Heating Fuel\n(HTFL) Height\n(HGHT)  \
0                  0.0  2 - FRME 2X4         3 - GAS                   NaN   
1                  0.0  3 - FRME 2X6         3 - GAS                   NaN   
2                  0.0  2 - FRME 2X4         3 - GAS                   NaN   
3                  0.0  2 - FRME 2X4         3 - GAS                   NaN   
4                  0.0  2 - FRME 2X4         3 - GAS                   NaN   

  Interior Wall\n(INT)             Neighborhood\n(NBHD)  \
0       0 - N/A                    1000200 - 4J MH PARK   
1       1 - DRYWALL                1000200 - 4J MH PARK   
2       2 - PLSTR/LTH   1000100 - SWISS VILLAGE MH PARK   
3       1 - DRYWALL     1000100 - SWISS VILLAGE MH PARK   
4       0 - N/A                    1000200 - 4J MH PARK   

   OPP - Open Porch\n(SubArea_OPP)  PTO - Patio\n(SubArea_PTO) Parcel Number  \
0                            176.0                         0.0  MOBILEM00034   
1                            160.0                         0.0  MOBILEM00037   
2                            104.0                         0.0  MOBILEM00040   
3                            212.0                         0.0  MOBILEM00042   
4                             24.0                         0.0  MOBILEM00043   

  Quality\n(QUAL)  RMS\n(RMS) Roof Cover\n(RCVR) Roof Structure\n(RSTR)  \
0  1 - LOW                0.0     3 - METAL              4 - ARCH         
1  2 - FAIR               0.0     3 - METAL              3 - GABLE        
2  2 - FAIR               NaN     3 - METAL              3 - GABLE        
3  2 - FAIR               0.0     3 - METAL              3 - GABLE        
4  1 - LOW                0.0     3 - METAL              3 - GABLE        

          Type  WBL - Wood Balcony\n(SubArea_WBL)  
0  Mobile Home                                0.0  
1  Mobile Home                                0.0  
2  Mobile Home                                0.0  
3  Mobile Home                                0.0  
4  Mobile Home           

#### Here is what needs to be done to the current dataframe (as of 9/24/24):

**Drop Row**
- 'Actual Year Built\n(AYB)'
- 'Neighborhood\n(NBHD)' drop row if there are not a lot of NaN values. If there are a lot of NaN values, I may just have to drop this column, because I think the only alternative if spatially locating the parcel with the neighborhood. That would require getting the neighborhood shapefiles from online or from the tax assessor.

**Fill with constant value**
- 'Air Conditioning\n(AIRC)' fill with '0 - N/A' but make sure there are no spaces in that value for the column
- 'Exterior Percent\n(EXW_PERCENT)' fill with 100
- 'Roof Cover\n(RCVR)' will with '0 - N/A'
- 'Roof Structure\n(RSTR)' fill with '0 - N/A'. However, there are a lot of spaces within the strings, so I will have to resolve that before doing anything.


**Fill with most common value in neighborhood**
- 'Architecture Style\n(ARCH)'
- 'Exterior Wall\n(EXW)'
- 'Heating Fuel\n(HTFL)'

**Fill with average value in the neighborhood**
- 'Effective Year Built\n(EYB)' if there are a lot of NaN values I will just drop the column. If there are not a lot, I will take the average value of the neighborhood.
- 'Height\n(HGHT)'

**Fill with most common value in dataframe**
- 'Floor\n(FLR)'
- 'Frame\n(FRME)'
- 'Interior Wall\n(INT)'

**Fill with average value in the dataframe**
- 'RMS\n(RMS)'

**More work to be done**
- 'Condition\n(COND)' contains a lot of redundancy in values that will have to be cleaned up. NaN will be filled with the median i.e. '4 - AVERAGE'

In [142]:
# Use this code to iterate through columns and see unique values to understand how to clean up data
print(df['Roof Structure\n(RSTR)'].unique())

['4 - ARCH      ' '3 - GABLE     ' '1 - FLAT      ' '2 - SHED      '
 '6 - GABEL MED ' '5 - GABEL LOW ' '8 - MULTI PEAK' '11 - HIP       '
 '3 - SHED MED  ' '7 - GABEL STP ' '0 - N/A       ' '4 - HIP       '
 '6 - MANSARD   ' '9 - MANSARD   ' '4 - SHED STP  ' '2 - SHED LOW  '
 '7 - GAMBREL   ' '10 - GAMBREL   ' '12 - BOWSTRING ' '8 - ARCH      ' nan
 '14 - GEO DOME  ' '13 - DOME      ']
